api/provider/aws/deploy_create.go

```diff
@@ -47,6 +47,46 @@ func (d *DeployProvider) createTaskDefinition(taskDefinition *ecs.TaskDefinition
 		input.SetNetworkMode(nm)
 	}
 
+	// We'll be explicit here and set any and all compatibilities that a user specifies in the task definition.
+	// At the moment, that should only be "FARGATE" and "EC2".
+	requiresCompatibilities := []*string{}
+	for _, compatibility := range taskDefinition.RequiresCompatibilities {
+		requiresCompatibilities = append(requiresCompatibilities, compatibility)
+
+		// There are some additional requirements for a task definition to be considered Fargate-compatible:
+		// https://github.com/aws/aws-sdk-go/blob/master/service/ecs/api.go#L8172
+		// https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definition_parameters.html#task_size
+		//
+		// For the most part, there are four things that should be specified in a Fargate task definition
+		// that aren't required for an EC2 task definition:
+		//   "requiresCompatibilities": [ "FARGATE" ],
+		//   "networkMode": "awsvpc",
+		//   "cpu": "<an appropriate value>",
+		//   "memory": "<an appropriate value>"
+		if aws.StringValue(compatibility) == ecs.LaunchTypeFargate {
+			cpu := aws.StringValue(taskDefinition.Cpu)
+			memory := aws.StringValue(taskDefinition.Memory)
```

Comment:
so there is memory and memory reservation; if memory is not set, should we check for memory reservation and use that if it is set?

Reply:
`memoryReservation` is a container-level parameter, not a task-definition-level parameter. Specifying `memoryReservation` at the task-definition level at all should be an error, just like any other parameter that doesn't exist.

If you wanted to get fancy, we could inspect the container definitions, compile all the `memory` and `memoryReservation` specs, and do some maths to figure out what a proper `memory` value should be, but that seems like a lot of hidden magic and I'd prefer to avoid it.

Reply:
+1 for avoiding magic w `memory` and `memoryReservation`. AWS should throw an error if either of the params are left out. We should force the users to be explicit with those values.

Reply:
It should probably be noted that there exists a `memory` parameter at _both_ the task-definition-and container-definition-levels. Here's a summary of the relevant parts of the documentation at https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definition_parameters.html#task_size via a sample task definition file:

```
// task definition file
{
  // `memory` at the task-definition-level (TDL):
  // - is a string
  // - is required for Fargate instances, but optional for EC2 instances
  // - is coupled to the TDL `cpu` parameter in terms of valid values
  "memory": "512",
  "cpu": "256",
  "containerDefinitions": [
    {
      // `memory` at the container-definition-level (CDL):
      // - is an int
      // - is optional for both Fargate and EC2 instances, provided that
      //   at least one of `memory` and `memoryReservation` is supplied;
      //   if both are supplied, `memory` must be greater than
      //   `memoryReservation`.
      "memory": 256,
      "memoryReservation": 128,
    }
  ]
}
```

In terms of proactive inspection, `deploy_create.go` currently only checks that neither TDL-`memory` nor TDL-`cpu` are omitted. Given the summary above, do we want to do more? I could definitely write a more robust inspector that validates task- and container-definitions to those commented specifications, and while that will duplicate some of AWS' validation, it would also provide feedback to the user at the `deploy` step instead of later at the `service` step (which was the point of validating TDL-`memory` and -`cpu`).

Reply:
We should not couple `deploy`s to `launchType`s, which I think we would invariably end up doing if we add extra 'helper' logic to `deploy create`. 

Aside from that, my vote is to let AWS handle the errors and leave it at that. I know that is more minimalistic than we do in other areas of the code, but given how new Fargate is and how many changes I can foresee it having in the future, it seems like a lot less maintenance to defer to AWS for errors and not try to be too clever in advance.